scala3-tree-lifts/macro/src/main/scala/org/scalameta/adt/TreeLiftsGenerate.scala

```diff
@@ -0,0 +1,240 @@
+      val defaultBody: String = customMatcher(adt, defName, localName).getOrElse {
+        val init = "_root_"
+        def getNamePath(parts: Iterable[String]) = parts.foldLeft(init) { (acc, part) =>
+          s"$acc.$part"
+        }
+        val nameParts = adt.sym.fullName.split('.')
+        if (adt.sym.isClass) {
+          val fields = adt match { case leaf: Leaf => leaf.fields(isPrivateOK); case _ => Nil }
+          val args = fields.map(f => s"term($localName.${f.name})")
+          val latestAfterVersion =
+            if (adt.sym.isAstClass) {
+              val moduleNames = adt.sym.companion.info.decls.flatMap { x =>
+                if (x.isModule) Some(x.name.toString) else None
+              }
+              val latestAfterVersion = AstNamerMacros.getLatestAfterName(moduleNames).getOrElse {
+                c.abort(c.enclosingPosition, s"no latest version ${adt.sym.fullName}: $moduleNames")
+              }
+              latestAfterVersion :: Nil
+            } else Nil
+          val namePath = getNamePath(nameParts ++ latestAfterVersion)
+          val formattedArgs = (List(s"term($localName.origin)") ++ args).mkString(", ")
+          s"treeByMode('{$namePath}, $formattedArgs)"
+        } else getNamePath(nameParts)
+      }
+      val body = customWrapper(adt, defName, localName, defaultBody).getOrElse(defaultBody)
+      s"def $defName($localName: ${adt.tpe}) = $body"
+    }
+    val clauses = adts.zip(defNames).map { case (adt, name) =>
+      s"case (y @ (_: ${adt.tpe})) => $name(y.asInstanceOf[${adt.tpe}])"
+    }
+
+    val retStr =
+      s"""|
+          |package scala.meta
+          |package internal
+          |package quasiquotes
+          |
+          |import scala.runtime.ScalaRunTime
+          |import scala.quoted._
+          |import scala.meta.{Tree => MetaTree}
+          |import scala.meta.internal.trees.Quasi
+          |import scala.meta.internal.parsers.Absolutize._
+          |import scala.meta.internal.parsers.Messages
+          |
+          |import scala.collection.mutable
+          |import scala.annotation.tailrec
+          |import scala.meta.trees.Origin
+          |
+          |$treeLiftsTrait
+          |
+          |trait TreeLifts(using override val internalQuotes: Quotes)(isPatternMode: Boolean, dialectExpr: Expr[Dialect]) extends ReificationMacros with TreeLiftsTrait {
+          |  import internalQuotes.reflect._
+          |$treeByMode
+          |
+          |$termMethods
+          |
+          |  ${liftAdts.mkString("\n  ")}
+          |
+          |  def liftableSubTree0[T <: MetaTree](y: T)(using Quotes): Tree = {
+          |    y match {
+          |      ${clauses.mkString("\n      ")}
+          |      case _ => sys.error("none of leafs matched " + (y.getClass))
+          |    }
+          |  }
+          |}
+          |""".stripMargin
+    val retStrList = Literal(Constant(retStr))
+    c.Expr[String](retStrList)
```

Comment:
is this a way to turn a string into a compiled code in scala3?

Reply:
We return the String in a scala 2 macro so that the runtime code (scala3-tree-lifts/impl/src/main/scala/org/scalameta/adt/Main.scala) can generate the scala 3 file.